[quant][fx] Embedding quantization support

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46678 [quant][fx] EmbeddingBag quantization support
* **#46677 [quant][fx] Embedding quantization support**

Summary:
Add support for weight only embedding quantization

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_qembedding_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24463305](https://our.internmc.facebook.com/intern/diff/D24463305)

supriyar committed:
[quant][fx] Embedding quantization support
Summary:
Add support for weight only embedding quantization

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_qembedding_module

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -378,6 +378,27 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
+    def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_dict=None):
+        qconfig = quantizer.qconfig_map[node.name]
+        assert not activation_is_statically_quantized(qconfig)
+        qemb = nnq.Embedding
```
isn't this dynamic quant? why is this nnq instead of nnqd?

supriyar replied:
It is weight only quantization. Main use case for embedding quantization is recommender models which use static quant to quantize embeddings and FC (mainly). So we decided to keep this in nnq for ease of use.

jerryzh168 replied:
I see, now that we changed api in fx, I feel we should probably put this in a separate namespace to make it clear that it is weight only quantization
e.g. `nn.quantized.weight_only`

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -378,6 +378,27 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
+class Embedding(QuantizeHandler):
+    def __init__(self, quantizer, node):
+        super().__init__(quantizer, node)
+        assert node.op == 'call_module'
+        self.emb_node = node
+        self.emb = quantizer.modules[self.emb_node.target]
```
these can happen in convert I think

supriyar replied:
I see. Was following convention of other patterns in this file. Is there a benefit of moving this to convert?

jerryzh168 replied:
some of the other patterns need this because there are optional nodes in the pattern I think.
I think it's simpler to move this to convert, it's not necessary. to do this in __init__ in this case.

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -378,6 +378,27 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
+        return quantizer.quantized_graph.create_node(
+            'call_module',
+            self.emb_node.target,
+            (load_arg(quantized=False)(self.emb_node.args[0]),), {})
```
since we don't quantize anything I think it would be better to use
```
return quantizer.quantized_graph.create_node(
            'call_module',
            self.emb_node.target,
            load_arg(quantized=False)(self.emb_node.args), load_arg(quantized=False)(self.emb_node.kwargs))
```

and later we might be able to generalize this to other wight only quant ops.

jerryzh168 commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -604,7 +604,8 @@ def checkGraphModeFxOp(self, model, inputs, quant_type,
                            debug=False,
-                           print_debug_info=False):
+                           print_debug_info=False,
+                           qconfig_input=None):
```
nit: qconfig_input -> custom_qconfig

jerryzh168 reviewed (approved):
Looks good, thanks

supriyar committed:
Update on "[quant][fx] Embedding quantization support"


Summary:
Add support for weight only embedding quantization

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_qembedding_module

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -378,6 +378,28 @@ def convert(self, quantizer, node, load_arg, debug=False, convert_custom_config_
 
+@register_quant_pattern(torch.nn.Embedding)
+class Embedding(QuantizeHandler):
+    def __init__(self, quantizer, node):
```
nit: this can be removed

supriyar committed:
Update on "[quant][fx] Embedding quantization support"


Summary:
Add support for weight only embedding quantization

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_qembedding_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24463305](https://our.internmc.facebook.com/intern/diff/D24463305)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Embedding quantization support"


Summary:
Add support for weight only embedding quantization

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_qembedding_module

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24463305](https://our.internmc.facebook.com/intern/diff/D24463305)

[ghstack-poisoned]
